File: util/events/emitter.go
package events

import (
	"context"
	"github.com/RomanIschenko/notify"
	"github.com/RomanIschenko/notify/pubsub"
	"reflect"
	"sync"
)

// Emitter helps you send and receive different events
// The event encoding algorithm:
// 1) 1st byte is a length of event name
// 2) from second to 1+<NAME_LENGTH> is the name of event
// 3) other bytes are json encoded data
type Emitter struct {
	app *notify.App
	handlers map[string]*handler
	codec Codec
	rv reflect.Value
	mu sync.RWMutex
}

func (e *Emitter) Emit(ev Event) {
	bts, err := e.codec.Marshal(ev.Data)
	if err != nil {
		return
	}

	data := make([]byte, 0, 1+len(ev.Name)+len(bts))
	data = append(data, byte(len(ev.Name)))
	data = append(data, []byte(ev.Name)...)
	data = append(data, bts...)

	e.app.Publish(pubsub.PublishOptions{
		Topics:      ev.Topics,
		Clients:     ev.Clients,
		Users:       ev.Users,
		Message:     data,
		NoBuffering: ev.NoBuffering,
		MetaInfo:    ev.MetaInfo,
		Seq:         ev.Seq,
	})
}

func (e *Emitter) On(name string, handler interface{}) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.handlers[name] = newHandler(handler, e.codec)
}

func (e *Emitter) processIncomingEvent(a *notify.App, data notify.IncomingData) {
	name, jsonEvent, err := parseIncomingData(data.Payload)
	if err != nil {
		logger.Errorln(err)
		return
	}
	e.mu.RLock()
	hnd, ok := e.handlers[name]
	e.mu.RUnlock()
	if !ok {
		return
	}
	hnd.call(reflect.ValueOf(e.app), e.rv, data.Client, jsonEvent)
}

// Creates new instance of emitter. If codec is nil json codec is used
func NewEmitter(ctx context.Context, app *notify.App, codec Codec) *Emitter {
	if codec == nil {
		codec = JSONCodec{}
	}
	emitter := &Emitter{
		app:      app,
		handlers: map[string]*handler{},
		mu:       sync.RWMutex{},
		codec: codec,
	}
	emitter.rv = reflect.ValueOf(emitter)
	app.Events(ctx).OnIncomingData(emitter.processIncomingEvent)
	return emitter
}

File: examples/xmas-mafia/client/lib/src/emitter.js
export const JSONCodec = {
    marshal(data) {
        return JSON.stringify(data)
    },
    unmarshal(data) {
        return JSON.parse(data)
    }
}

export class Emitter {
    constructor(client, codec=JSONCodec) {
        this.__client = client
        this.__handlers = {}
        this.__codec = codec
        this.__client.on('message', data => {
            try {
                let event = this.__parseEvent(data)
                let handlers = this.__handlers[event.name]
                if(handlers) {
                    for(let h of handlers) h(event.data)
                }
            } catch(e) {
                console.error('error while processing the incoming event:', e)
            }
        })
    }

    __parseEvent(payload) {
        if(payload.length < 2) {
            throw new Error('received message too short')
        }
        let nameLength = payload[0].charCodeAt(0)
        if(payload.length < nameLength + 1) {
            throw new Error("name length is too big")
        }
        let name = payload.slice(1, 1+nameLength)
        let data = this.__codec.unmarshal(payload.slice(1+nameLength))
        return {name, data}
    }

    off(event, handler) {
        let eventHandlers = this.__handlers[event]
        if(!eventHandlers) {
            return
        }
        let i = eventHandlers.indexOf(handler)
        while(i >= 0) {
            eventHandlers[i] = eventHandlers[eventHandlers.length-1]
            i = eventHandlers.indexOf(handler)
        }
        if(eventHandlers.length === 0) {
            delete this.__handlers[event]
        }
    }

    on(event, handler) {
        let eventHandlers = this.__handlers[event]
        if(!eventHandlers) {
            eventHandlers = []
            this.__handlers[event] = eventHandlers
        }
        eventHandlers.push(handler)
    }

    emit(event, data) {
        try {
            let marshaledData = this.__codec.marshal(data)
            let eventLength = event.length
            if(eventLength > 255) {
                console.error('event length is bigger than 255')
                return
            }
            this.__client.send(String.fromCharCode(eventLength)+event+marshaledData)
        } catch(e) {
            console.error("error while emitting:", e)
        }
    }
}